enhancements/big-bundle-unpacking.md

```diff
@@ -0,0 +1,138 @@
+
+### Implementation Details/Notes/Constraints
+
+#### PodSpec Pull Requirement
+
+In order to support pulling bundle images on OpenShift, OLM must adhere to the platform's [pull policies](https://docs.openshift.com/container-platform/4.5/openshift_images/image-configuration.html). The lowest risk way to do this is to rely on CRI to pull images and apply the policies opaquely, and the most straightforward initiate a pull is via a PodSpec. This means that, when a bundle image is pulled and unpacked by OLM today, a `Job` is created which specifies three containers:
+
+- a bundle util init container
+  - contains a statically linked bundle tool that can search the container’s filesystem for bundle metadata -- annotations.yaml -- and copy the referenced manifest files to a bundle volume
+  - mounts a single bundle util volume
+  - at runtime, uses cp to copy this bundle tool to the bundle util volume
+- a bundle data init container
+  - pull spec is used to delegate pulling of the bundle image via CRI-O
+  - mounts two volumes: bundle util and bundle data
+  - at runtime, executes the bundle tool from the bundle util mount to copy the bundle to the bundle data mount
+- a bundle extract container
+  - contains opm
+  - mounts the bundle data volume
+  - runs `opm extract` against the bundle data mount which results in a `ConfigMap` being created that contains the bundle’s kubectl-able content
+
+#### Partitioning and Storage
+
+Instead of extracting bundle content directly to a `ConfigMap` -- as outlined in the [previous section](#podspec-pull-requirement) -- the unpack `Job` will first encode it as JSON and compress the result. The compressed JSON will then be split into chunks of a [fixed size](#determining-chunk-size), each wrapped with a header to keep track of its position in the chunk sequence. Each chunk is then encoded as JSON and stored in a respective `ConfigMap`. To each of these `ConfigMaps` a `porcelain.operators.coreos.com/bundle: <bundle-name>` label is applied to faciliate later retrieval. A `porcelain.operators.coreos.com/signature: <signature>` annotation is also generated for each chunk and is used to validate its integrity. When the content for a particular bundle is then retrieved, the `ConfigMaps` with the matching bundle name label are collected, sorted by the position tracked in each header, reconstituted into a single artifact, and decompressed before being returned for a `/content` request. If the signature verification fails for any chunk, a status condition indicating will be added to the associated bundle which will trigger a new unpack `Job` to run in a bid to rectify the issue.
+
+#### Determining Chunk Size
+
+Chunk-size will be determined by an option on the `catalog-operator` command, which if unset, directs the operator to probe the cluster for the resource storage limit. One potential probing strategy could be to create ConfigMaps of various sizes; akin to a "binary search" for the limit.
```

Comment:
I would leave the size probing as future work - there doesn't seem to be an immediate need, and it may be worth pursuing more direct options (i.e. should this value be exposed as a property of the cluster than anyone can query from kube-apisever?)